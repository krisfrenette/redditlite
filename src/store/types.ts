import {ActionNames} from 'store/action-names';

export type SubredditState = {
    name: string;
    posts: RedditPosts;
    before: string;
    after: string;
    count: number;
    currentPaging: {
        direction: PagingDirection;
        count: number;
    };
    fetchingData: boolean;
};

export interface IRootState {
    subreddit: SubredditState;
}

export type RedditPost = {
    id: string;
    author: string;
    title: string;
    url: string;
    created: number;
    commentUrl: string;
    thumbnail?: Thumbnail;
};

export type Thumbnail = {
    url: string;
    width: number;
    height: number;
};

export type PagingDirection = {
    direction: 'before' | 'after';
    value: string;
};

export type RedditPosts = RedditPost[];

export type SubredditActions = {
    FETCH_NEW_POSTS_REQUEST: {
        type: typeof ActionNames.FETCH_NEW_POSTS_REQUEST;
    };
    FETCH_NEW_POSTS_SUCCESS: {
        type: typeof ActionNames.FETCH_NEW_POSTS_SUCCESS;
        payload: {
            before: string;
            after: string;
            posts: RedditPosts;
            currentPaging: {
                direction: PagingDirection;
                count: number;
            };
        };
    };
    RESET_STATE: {
        type: typeof ActionNames.RESET_STATE;
    };
    FETCH_NEW_POSTS_FAILED: {
        type: typeof ActionNames.FETCH_NEW_POSTS_FAILED;
    };
};

export type RootAction = SubredditActions[keyof SubredditActions];
