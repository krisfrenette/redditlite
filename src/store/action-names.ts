export enum ActionNames {
    FETCH_NEW_POSTS_REQUEST = 'subreddit/fetch/new/request',
    FETCH_NEW_POSTS_SUCCESS = 'subreddit/fetch/new/success',
    FETCH_NEW_POSTS_FAILED = 'subreddit/fetch/new/failed',
    RESET_STATE = 'subreddit/reset'
}
