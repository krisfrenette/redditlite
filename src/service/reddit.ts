import * as URI from 'urijs';
import * as _ from 'lodash';
import {PagingDirection} from 'store/types';

type NewPostsResponse = {
    before: string;
    after: string;
    children: [];
};

const BASE_URL = 'https://www.reddit.com';

enum Uri {
    NewPosts = '/r/${subreddit}/new.json'
}

type NewPostsArgs = {
    subreddit: string;
    direction?: PagingDirection;
    limit: number;
    count: number;
};

type PathArgs = {uri: typeof Uri.NewPosts; subreddit: string};

export async function newPosts({
    subreddit,
    direction,
    limit,
    count
}: NewPostsArgs): Promise<NewPostsResponse> {
    const url = new URI(BASE_URL)
        .segment(getPath({uri: Uri.NewPosts, subreddit}))
        .addQuery({limit})
        .addQuery({count});

    if (direction && direction.direction) {
        url.addQuery(direction.direction, direction.value);
    }

    try {
        const response = await fetch(url.toString());
        if (response.status !== 200) {
            return null;
        }
        const {
            data: {children, before, after}
        } = await response.json();

        return {children, before, after};
    } catch (e) {
        return null;
    }
}

const getPath = ({uri, ...args}: PathArgs): string => {
    if (args) {
        const compiled = _.template(uri);
        return compiled(args);
    } else {
        return uri;
    }
};
