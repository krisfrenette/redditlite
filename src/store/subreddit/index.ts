import {SubredditState, RootAction} from 'store/types';
import {ActionNames} from 'store/action-names';

const getDefaultState = (): SubredditState => {
    return {
        name: null,
        posts: [],
        before: null,
        after: null,
        count: 0,
        currentPaging: null,
        fetchingData: true
    };
};

export default function reducer(
    state: SubredditState = getDefaultState(),
    action: RootAction
): SubredditState {
    switch (action.type) {
        case ActionNames.FETCH_NEW_POSTS_REQUEST:
            return {
                ...state,
                fetchingData: true
            };
        case ActionNames.FETCH_NEW_POSTS_SUCCESS:
            return {
                ...state,
                fetchingData: false,
                ...action.payload
            };
        case ActionNames.FETCH_NEW_POSTS_FAILED:
            return {
                ...state,
                fetchingData: false
            };
        case ActionNames.RESET_STATE:
            return getDefaultState();

        default:
            return state;
    }
}
