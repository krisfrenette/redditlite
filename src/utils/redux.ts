import {IRootState, RootAction} from 'store/types';
import {ThunkAction} from 'redux-thunk';

// Convenience type for redux thunks
export type Thunk<R = void, S = IRootState> = ThunkAction<
    R,
    S,
    void,
    RootAction
>;
