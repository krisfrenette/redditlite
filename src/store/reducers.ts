import {combineReducers} from 'redux';
import subreddit from 'store/subreddit';

export default combineReducers({
    subreddit
});
