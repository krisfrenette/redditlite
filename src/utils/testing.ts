import * as _ from 'lodash';
import {Thunk} from './redux';
import {IRootState} from 'store/types';

export const mockDispatch = (getState): jest.Mock => {
    const dispatch = jest.fn(async (action: Thunk) => {
        if (_.isFunction(action)) {
            return await action(dispatch, getState, undefined);
        }
    });

    return dispatch;
};

export const mockGetState = (
    state: Partial<IRootState> = {}
) => (): IRootState => state as any;
