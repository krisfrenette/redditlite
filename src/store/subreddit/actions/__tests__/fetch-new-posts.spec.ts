// Fix for Node clobbering the global variable
const globalAny: any = global;

import {Thunk} from 'utils/redux';
import {
    fetchNewPosts,
    DEFAULT_LIMIT
} from 'store/subreddit/actions/fetch-new-posts';
import {mockDispatch, mockGetState} from 'utils/testing';
import {ActionNames} from 'store/action-names';
import {newPosts} from 'service/reddit';

let mockShouldFail = false;

const mockCurrentPaging = {
    direction: {
        direction: 'after',
        value: 'mockCurrentPagingValue'
    },
    count: 100
};

const scrollToSpy = jest.fn();
globalAny.scrollTo = scrollToSpy;

jest.mock('service/reddit', () => ({
    newPosts: jest.fn(() => {
        if (mockShouldFail) {
            return null;
        }
        return {
            before: 'before',
            after: 'after',
            children: [
                {
                    data: {
                        author: 'mockAuthor1',
                        permalink: 'mockPermalink1',
                        created: 12345,
                        id: 'mockId1',
                        thumbnail: null,
                        thumbnail_height: null,
                        thumbnail_width: null,
                        title: 'mockTitle1',
                        url: 'mockUrl1'
                    }
                }
            ]
        };
    })
}));

const mockCount = 50;

const getState = mockGetState({
    subreddit: {
        count: mockCount,
        currentPaging: mockCurrentPaging
    }
} as any);
const dispatch = mockDispatch(getState);
const mockSubreddit = 'mockSubreddit';

beforeEach(() => {
    mockShouldFail = false;
    jest.clearAllMocks();
});

describe('fetchNewPosts', () => {
    const performCall = payload => {
        const thunk: Thunk = fetchNewPosts(payload);
        return thunk(dispatch, getState, undefined);
    };

    it('fetch calls request for a given subreddit', async () => {
        await performCall({subreddit: mockSubreddit});
        expect(dispatch).toBeCalledWith({
            type: ActionNames.FETCH_NEW_POSTS_REQUEST
        });
    });

    it('fetch successfully gets posts for a given subreddit', async () => {
        await performCall({subreddit: mockSubreddit});
        expect(dispatch).toMatchSnapshot();
    });

    it('sends a failed action when no posts are returned', async () => {
        mockShouldFail = true;
        try {
            await performCall({subreddit: mockSubreddit});
        } catch (e) {}
        expect(dispatch).toBeCalledWith({
            type: ActionNames.FETCH_NEW_POSTS_FAILED
        });
    });

    it('calls newPosts', async () => {
        await performCall({subreddit: mockSubreddit});
        expect(newPosts).toMatchSnapshot();
    });

    it('fetch calls with after', async () => {
        const mockDirection = {direction: 'after', value: 'mockValue'};

        await performCall({
            subreddit: mockSubreddit,
            direction: mockDirection,
            limit: DEFAULT_LIMIT
        });
        expect(newPosts).toBeCalledWith({
            subreddit: mockSubreddit,
            limit: DEFAULT_LIMIT,
            direction: mockDirection,
            count: mockCount + DEFAULT_LIMIT
        });
    });

    it('fetch calls scroll to top when scrollTop is true', async () => {
        const mockDirection = {direction: 'after', value: 'mockValue'};

        await performCall({
            subreddit: mockSubreddit,
            direction: mockDirection,
            limit: DEFAULT_LIMIT,
            scrollTop: true
        });
        expect(scrollToSpy).toBeCalledWith(0, 0);
    });

    it('fetch calls with before', async () => {
        const mockDirection = {direction: 'before', value: 'mockValue'};

        await performCall({
            subreddit: mockSubreddit,
            direction: mockDirection,
            limit: DEFAULT_LIMIT
        });
        expect(newPosts).toBeCalledWith({
            subreddit: mockSubreddit,
            limit: DEFAULT_LIMIT,
            direction: mockDirection,
            count: mockCount
        });
    });

    it('fetch calls refresh data', async () => {
        const mockDirection = {direction: 'before', value: 'mockValue'};

        await performCall({
            subreddit: mockSubreddit,
            direction: mockDirection,
            refreshData: true,
            limit: DEFAULT_LIMIT
        });
        expect(newPosts).toBeCalledWith({
            subreddit: mockSubreddit,
            limit: DEFAULT_LIMIT,
            direction: mockCurrentPaging.direction,
            count: mockCurrentPaging.count
        });
    });
});
