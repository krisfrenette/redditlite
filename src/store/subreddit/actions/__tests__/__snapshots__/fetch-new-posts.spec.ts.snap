// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchNewPosts calls newPosts 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "count": 50,
        "direction": null,
        "limit": 25,
        "subreddit": "mockSubreddit",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Object {
        "after": "after",
        "before": "before",
        "children": Array [
          Object {
            "data": Object {
              "author": "mockAuthor1",
              "created": 12345,
              "id": "mockId1",
              "permalink": "mockPermalink1",
              "thumbnail": null,
              "thumbnail_height": null,
              "thumbnail_width": null,
              "title": "mockTitle1",
              "url": "mockUrl1",
            },
          },
        ],
      },
    },
  ],
}
`;

exports[`fetchNewPosts fetch successfully gets posts for a given subreddit 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "type": "subreddit/fetch/new/request",
      },
    ],
    Array [
      Object {
        "payload": Object {
          "after": "after",
          "before": "before",
          "count": 50,
          "currentPaging": Object {
            "count": 50,
            "direction": null,
          },
          "posts": Array [
            Object {
              "author": "mockAuthor1",
              "commentUrl": "mockPermalink1",
              "created": 12345,
              "id": "mockId1",
              "thumbnail": Object {
                "height": null,
                "url": null,
                "width": null,
              },
              "title": "mockTitle1",
              "url": "mockUrl1",
            },
          ],
        },
        "type": "subreddit/fetch/new/success",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;
